File: Classic-Algorithms/HeapsAlgorithm.py

def heaps_algorithm(arr):

    output = []

    def swap(swap_arr, i, j):
        t = swap_arr[i]
        swap_arr[i] = swap_arr[j]
        swap_arr[j] = swap_arr[i]
    
    def generate(heap_arr, length):
        if length == 0:
            output.append(heap_arr[:])
            return

        generate(heap_arr, length - 1)
        for i in range(length - 1):
            if length % 2 == 0:
                swap(heap_arr, i, length - 1)
            else:
                swap(heap_arr, 0, length - 1)
            generate(heap_arr, length - 1)
    
    generate(arr[:], len(arr))
    return output
        

File: Classic-Algorithms/HeapsAlgorithm.js

function getPermutations(arr) {
    const output = []
    function swap(arrToSwap, i, j) {
        const t = arrToSwap[i]
        arrToSwap[i] = arrToSwap[j]
        arrToSwap[j] = t
    }
    function generate(len, heapArr) {
        if (len === 1) {
            output.push(heapArr.slice())
            return
        }
        generate(len - 1, heapArr) 
        for (let i = 0; i < len - 1; i++) {
            if (len % 2 == 0) {
                swap(heapArr, i, len - 1)
            } else {
                swap(heapArr, 0, len - 1)
            }
            generate(len - 1, heapArr)
        }
    }
    generate(arr.length, arr.slice())
    return output
}

const test = [1,2,3]
console.log(getPermutations(test))
